### PR TITLE
feat: add advance search and collection

### PR DESCRIPTION
**Pull request by kilosonc**

### CODE REVIEW

File: src/pages/dashboard/Applications/index.tsx [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
 import { listApplications } from '@/services/applications/applications';
 import '@/components/GroupTree/index.less';
-import type { API } from '@/services/typings';
 import { DTree } from '@/components/DirectoryTree';
 import { ComponentWithPagination, PageWithInitialState, PageWithInitialStateProps } from '@/components/Enhancement';
-import withTrim from '@/components/WithTrim';
-import TitleWithCount from '../components/TitleWithCount';
 import WithContainer from '../components/WithContainer';
 import { setQuery } from '../utils';
+import HorizonAutoCompleteHandler, { AutoCompleteOption } from '../../../components/FilterBox/HorizonAutoCompleteHandler';
```
Comment (xuzhu-591) [inline src/pages/dashboard/Applications/index.tsx]: How about using `@/components/...`

File: src/pages/dashboard/Applications/index.tsx [hunk 2]
```diff
@@ -31,16 +32,36 @@ interface MyApplicationsProps
 
 function Applications(props: MyApplicationsProps) {
   const { initialState, location } = props;
-  const { [QueryName]: qName = '', [QueryMode]: qMode = Mode.Own } = location.query ?? {};
 
   const intl = useIntl();
-  const [filter, setFilter] = useState(qName as string);
+  const [filter, setFilter] = useState<string>();
```
Comment (xuzhu-591) [inline src/pages/dashboard/Applications/index.tsx]: How about setting a default value?

Comment (xuzhu-591) [reply src/pages/dashboard/Applications/index.tsx]: The type is `string | undefined` now.

File: src/pages/dashboard/Clusters/index.less [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    color: #555;
+}
+
+
+.demo-4-filter-wrapper {
+    display: flex;
+    .demo-4-filter {
+        flex:1;
+    }
+}
```
Comment (xuzhu-591) [inline src/pages/dashboard/Clusters/index.less]: missing newline

File: src/pages/dashboard/Clusters/index.tsx [hunk 4]
```diff
@@ -183,137 +264,38 @@ function Clusters(props: ClustersProps) {
...
       ? initialState?.currentUser?.id
       : undefined,
     filter,
     pageSize,
     pageNumber,
     environment,
+    region: displayName2Region?.get(region),
     template: tpl,
     templateRelease: tplRelease,
+    isFavorite: (() => { if (isFavorite === 'true') return true; if (isFavorite === 'false') return false; return undefined; })(),
```
Comment (xuzhu-591) [inline src/pages/dashboard/Clusters/index.tsx]: The line is too long

File: src/pages/dashboard/Clusters/index.tsx [hunk 5]
```diff
@@ -331,44 +314,168 @@ function Clusters(props: ClustersProps) {
...
-        tabBarExtraContent={clusterQueryInput}
-        onChange={(key) => { setMode(key as Mode); }}
-        defaultActiveKey={mode}
-        animated={false}
-        style={{ marginTop: '15px' }}
+      <SearchBox
+        hKey="cluster"
+        autoCompleteHandler={handler}
+        defaultValue={defaultValue}
+        onSubmmit={onSubmit}
```
Comment (xuzhu-591) [inline src/pages/dashboard/Clusters/index.tsx]: onSubmmit => onSubmit

File: src/pages/dashboard/Clusters/index.tsx [hunk 6]
```diff
@@ -331,44 +314,168 @@ function Clusters(props: ClustersProps) {
...
+
+  const handler = useMemo(() => {
+    const options: AutoCompleteOption[] = [
+      {
+        key: 'env',
+        type: 'selection',
+        values: [
+          {
+            operator: '=',
+            possiableValues: envs ? envs.map((e) => e.name) : [],
```
Comment (xuzhu-591) [inline src/pages/dashboard/Clusters/index.tsx]: possibleValues

File: src/pages/dashboard/Clusters/index.tsx [hunk 7]
```diff
@@ -331,44 +314,168 @@ function Clusters(props: ClustersProps) {
...
+        values: [
+          {
+            operator: '=',
+            possiableValues: [],
+          },
+        ],
+        callback: (operator, trace) => {
+          for (let i = 0; i < trace.arr.length; i += 1) {
+            if (trace.arr[i].type === 'category' && trace.arr[i].value === 'template') {
+              const j = i + 2;
```
Comment (xuzhu-591) [inline src/pages/dashboard/Clusters/index.tsx:435]: Why this?

File: src/pages/dashboard/components/SearchBox.tsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    if (history === null) {
+      store.set(hKey, [result]);
+    } else {
+      if (result.length === 0) {
+        return;
+      }
+      history.push(result);
+      let newHistory: Expression[][] = [result, ...history];
+      const m: Record<string, null> = {};
+      newHistory = newHistory.filter((item) => { const h = hash(item); if (h in m) { return false; } m[h] = null; return true; });
```
Comment (xuzhu-591) [inline src/pages/dashboard/components/SearchBox.tsx]: The line is too long

File: src/services/clusters/clusters.ts [hunk 9]
```diff
@@ -276,3 +276,15 @@ export async function updateClusterTemplateSchemaTags(clusterID: number, data: A
     data,
   });
 }
+
+export async function addFavorite(clusterID: number) {
```
Comment (xuzhu-591) [inline src/services/clusters/clusters.ts:280]: why not use UpdateCluster ?

File: src/pages/dashboard/Applications/index.tsx (hunk 1 above)
Comment (kilosonc) [reply src/pages/dashboard/Applications/index.tsx]: fixed

File: src/pages/dashboard/Applications/index.tsx (hunk 2 above)
Comment (kilosonc) [reply src/pages/dashboard/Applications/index.tsx]: fixed

File: src/pages/dashboard/Clusters/index.less (hunk 3 above)
Comment (kilosonc) [reply src/pages/dashboard/Clusters/index.less]: fixed

File: src/pages/dashboard/Clusters/index.tsx (hunk 4 above)
Comment (kilosonc) [reply src/pages/dashboard/Clusters/index.tsx]: fixed

File: src/pages/dashboard/Clusters/index.tsx (hunk 5 above)
Comment (kilosonc) [reply src/pages/dashboard/Clusters/index.tsx]: fixed

File: src/pages/dashboard/Clusters/index.tsx (hunk 6 above)
Comment (kilosonc) [reply src/pages/dashboard/Clusters/index.tsx]: fixed

File: src/pages/dashboard/Clusters/index.tsx (hunk 7 above)
Comment (kilosonc) [reply src/pages/dashboard/Clusters/index.tsx:435]: retrive template release by template user chosed, j is index for key, j + 1 is index for operator, j + 2 is index for value

File: src/services/clusters/clusters.ts (hunk 9 above)
Comment (kilosonc) [reply src/services/clusters/clusters.ts:280]: UpdateCluster will update every field in a cluster, that's unnecessary.

**Review by xuzhu-591 — APPROVED** [review]

**Pull request merged**